pyface/ui/qt4/application_window.py

```diff
@@ -115,6 +115,28 @@ def _set_window_icon(self):
         if self.control is not None:
             self.control.setWindowIcon(icon.create_icon())
 
+    # ------------------------------------------------------------------------
+    # 'Widget' interface.
+    # ------------------------------------------------------------------------
+
+    def destroy(self):
+        if self.control is not None:
+            menus = self.control.findChildren(QtGui.QMenu)
+            for menu in menus:
+                try:
+                    menu.disconnect_event_listeners()
+                except AttributeError:
+                    continue
+
+            toolbars = self.control.findChildren(QtGui.QToolBar)
+            for toolbar in toolbars:
+                try:
+                    toolbar.disconnect_event_listeners()
+                except AttributeError:
+                    continue
+
+        super(ApplicationWindow, self).destroy()
+
```

Comment:
I agree this is ugly and is not very scalable. If we want to do this, we should do this right.
`pyface.ui.qt4.action.MenuManager.destroy` is called when the application window is destroyed. Currently that method is not implemented and calls the default implementation. I think we should implement that `destroy` method to do the clean up there.  The menu manager might need to hold references to the widget in order to clean up in `destroy`, but those references can be removed in the same `destroy` method. Likewise I think we can add clean up to the `destroy` method of the toolbar manager.

In this `ApplicationWindow`, there is a change handler for `@on_trait_change("tool_bar_manager, tool_bar_managers")`. I think the destroy method applies there too. That may be out-of-scope for this task, but we should open an issue for that. This perhaps supports the idea of implementing the `destroy` method as it would be more reusable in a different context.